onnxruntime/core/framework/sequential_executor.cc

```diff
@@ -125,8 +126,71 @@ static Status ReleaseNodeMLValues(ExecutionFrame& frame,
 Status SequentialExecutor::Execute(const SessionState& session_state, const std::vector<int>& feed_mlvalue_idxs,
                                    const std::vector<OrtValue>& feeds, const std::vector<int>& fetch_mlvalue_idxs,
                                    std::vector<OrtValue>& fetches,
-                                   const std::unordered_map<size_t, CustomAllocator>& fetch_allocators,
-                                   const logging::Logger& logger) {
+                                   const std::unordered_map<size_t, IExecutor::CustomAllocator>& fetch_allocators,
+                                   const logging::Logger& logger,
+                                   int64_t& run_id) {
+  const SequentialExecutionPlan& seq_exec_plan = *session_state.GetExecutionPlan();
+  const auto& exec_plan_vec = seq_exec_plan.execution_plan;
+  const auto exec_plan_size = exec_plan_vec.size();
+  if (run_id == DEFAULT_RUN_ID) {
+    ExecutionFrame frame{feed_mlvalue_idxs, feeds, fetch_mlvalue_idxs, fetches, fetch_allocators, session_state};
+    return Execute(session_state, feeds, fetch_mlvalue_idxs, fetches, logger, frame, 0, exec_plan_size);
+  } else {
+    // Partial graph execution frame management code.
+    ExecutionFrame* frame;
+    std::map<int64_t, std::unique_ptr<onnxruntime::ExecutionFrame>>::iterator it;
+    if (run_id == DEFAULT_PARTIAL_RUN_ID) {
+      auto p_frame = onnxruntime::make_unique<ExecutionFrame>(feed_mlvalue_idxs, feeds, fetch_mlvalue_idxs, fetches,
+                                                              fetch_allocators, session_state);
+
+      frame = p_frame.get();
+      std::lock_guard<OrtMutex> lock(session_state.graph_runs_lock_);
+      it = session_state.graph_runs_.insert(std::make_pair(session_state.graph_runs_counter_, std::move(p_frame))).first;
+      ORT_ENFORCE(run_id < INT64_MAX);
+      run_id = session_state.graph_runs_counter_;
+      // REVIEW(mzs): May be reuse counters and prevent a potential overflow?
+      session_state.graph_runs_counter_ += 1;
+    } else {
+      ORT_ENFORCE(run_id >= 0);
+
+      it = session_state.graph_runs_.find(run_id);
+
+      ORT_ENFORCE(it != session_state.graph_runs_.end());
+
+      (it->second)->UpdateFeedAndFetches(feed_mlvalue_idxs, feeds, const_cast<std::vector<int>&>(fetch_mlvalue_idxs), fetches);
+      frame = (it->second).get();
+    }
+
+    // Determine partial graph execution start and end nodes.
+    // REVIEW(codemzs): Ideally these should be determined just by looking at fetches but we can make that 
+    // optimization after the first commit. That will also get rid of the need for an "explicit" operator to indicate 
+    // partial graph execution cut point, this will require some thinking on making sure the control flow of execution order
+    // is correct, i.e backward pass follows forward pass.
```

Comment:
I agree with you that doing graph analysis or string munging at runtime is inefficient and that was _not_ my intent, my thinking was along the lines where we have hints in the graph as to which direction to continue (without needing an op) and executor will continue to execute until the desired tensors are realized. I _**do not**_ plan on implementing this solution in this commit but I will definitely try in my free time and run by you when you have some free cycles to get your feedback ... no plans to check it in unless you feel otherwise ... just for my own curiosity is all :) We definitely do not want to do redundant work that can hinder scalability. Thanks, @mrry  for reviewing my change and providing valuable feedback.